### ISSUE TITLE
Support for different timezones

### ISSUE THREAD

**Issue by mipapo** [Runalyze/Runalyze#1145]
Wie Myriam anmerkte. Zeitzonen sind noch ein großes Problem. Wie in #756, wodurch die Wetterkorrektur schon mal stimmt, aber die Zeitzone wird noch nicht angepasst oder täusche ich mich?

**Comment by laufhannes** [comment]
+1 via mail from Jeremy

> The Date/Time also appear to be importing funky. It looks like it may be assuming I'm in a different time zone than I really am. Is there a date adjustment field I'm missing.

**Comment by laufhannes** [comment]
- [ ] make sure that all timestamps are in UTC
 - [ ] add an account option 'timezone'
   - [ ] correctly set this value on registration
   - [ ] use the server timezone in our next update for existing accounts
 - [ ] display all dates/times based on the account's timezone
 - [ ] make sure that timezone-related queries are working properly

Further readings:
 - http://www.thisprogrammingthing.com/2013/things-i-learned-while-writing-a-timezone-aware-website/
 - http://stackoverflow.com/questions/10834665/how-can-i-handle-time-zones-in-my-webapp/10924961#10924961

**Comment by laufhannes** [comment]
What is if I'm in another timezone, e.g. for vacations, and add some activities there? They will be saved in UTC but displayed in my home's timezone, right?

Or even worse: What is if I move after five years of running from mid-Europe to the US? If I change my account's timezone, all previous activities will now be displayed as nightly activities, right?

---

I guess the only solution would be to save the timezone of each activity? (Which will complicate all time-related queries, e.g. to display all activities of a week)

**Comment by mipapo** [comment]
Save all activities in UTC. 
Add User setting: Timezone or based on location setting. 
If user uploads a file then check if a timezone information is available or fallback to user setting. 
Maybe for statistic reasons or other features (currently nothing occurs to me) add a column in _training where the original time zone is stored. (Maybe for exporting the activity?)

**Comment by grzeg1** [comment]
IMO each activity should be saved as it's local time+time zone. It should be always displayed as performed at it's local timezone. If I were in US for a week and trained at 9AM there, I should see 9AM in diary, regardless of when it's displayed.
User should be able to set his time zone in config to facilitate import. Time zone should be settable on import form.

The only time when timezone calculation is needed is calculating CTL/ATL/TSB. They could be calculated based on UTC and offset to current time zone. But also since they're calculated on a daily basis, the error would only matter if you were constantly travelling through several time zones, which is improbable.

Weekly/monthly/yearly reports are not that much important that they couldn't use activity's local time.

**Comment by mipapo** [comment]
@grzeg1 I am not sure whether I understood your opinion right. So save the activity in UTC and like I said add the timezone in an extra field. Or did you mean to save the timezone in a timestamp too? 

Where to you think should the time zone settable in the import form? Before I begin to upload files? 

Good hint (calculation CTL, ATL,TSB)

**Comment by grzeg1** [comment]
I'm not sure yet what exactly the MySQL's options are so maybe I rushed a little bit.
My idea was to stay with current way of saving activity times, just add another column with time zone offset. This column could be used for e.g CTL/ATL/TSB if needed.
My point is, if something happened 9AM local time, it should be displayed this way, regardless if it's my own diary or shared page. At least I wouldn't want my activities displayed in any other way.
Probably addressing 2 main problems with time zones: file import and determining what time is now/which day is today in user's location will suffice.
Training diary is not something like Facebook page where you want to see what time posts were added and need everything in time order.

**Comment by mipapo** [comment]
I see it exactly the same way. I had a look to other sports diaries and they all do it the same way.

Another point: e.g. for the online version and everything else the current activity times have to "converted" to UTC.

**Comment by grzeg1** [comment]
For online version/local installs: I would leave TZ offset as empty for all activities and ask user to select his/her time zone after first login, then update all activities to chosen TZ offset (for online version could be via cron script).
If it's too much work, I'd just leave these as 0 offset. Or maybe use the database clean menu and create "set tz offset to your configured offset" option (to be removed in the next version).

**Comment by mipapo** [comment]
to sum it up:
- current timestamp has to updated to UTC during update and after the first login the user chooses her/his TZ and the offset is updated for old activities
Correct?
Don't think we will add this in version 2.2 (FYI: development stop on sunday, afterwards only bugfixes)

**Comment by grzeg1** [comment]
@mipapo to be clear, my idea was NOT updating curent timestamps and NOT storing them in UTC but as local time

**Comment by mipapo** [comment]
Don't think that it's an good idea to store timestamps of activities of different users in local times. UTC would make it all easier. 
When users change in their setting their current timezone that would be rubbish.
It's easier to work with timestamp in UTC. TZ offset can be saved in an other field for any features we haven't thought of yet.
The timezone in the user settings would just change the functions e.g. in the databrowser "what day is today etc."

**Comment by laufhannes** [comment]
The clear advise from stackoverflow is to store everything in UTC.

Store in UTC | Store in local time
------------ | -------------
PHP- and MySQL-time zone should be set to UTC | PHP and MySQL need users' time zone
`date()` and `FROM_UNIXTIME()` are correct for every activity | `date()` and `FROM_UNIXTIME()` require the activity's time zone
relative times ('today', 'X hours ago') need user's time zone (which is different to PHP's/MySQL's one) | relative times ('today', 'X hours ago') need user's time zone (which is equal to PHP's/MySQL's one)
does work without time zones for activities | does require correct time zones for activities

 * ATL/CTL: I don't see a problem here. A few hours shift (which will occur seldom) is negligible.
 * Import: Most files use timestamps in UTC, otherwise `strtotime()` will correctly transform them. And any input as '07.10.2015, 10:00' regardless of the user's time zone if we use UTC.

The only need for user's time zone is to define `NOW`.

**Comment by grzeg1** [comment]
I think we're talking different things.

If I store 2015-10-07 10:25 (now) in database without time zone, when reading it's always going to be that same timestamp, regardless of which time zone I will be when reading. 
If that's activity start, that's what I want to display. Always. So no need for timezone info for activities here.

Now when I want to import data, files usually include dates in UTC. So for import I need timezone to convert data to local time and store it (just like it's done currently).

The only time we need activity's time zone is when I want to calculate CTL/ATL accurately.
I also need user's time zone to define `NOW`.

So in other words: add user's timezone to config to define `NOW`, add timezone selection to import dialog and we're good with 99% cases.

**Comment by grzeg1** [comment]
And StackOverflow advice is for multi-timezone apps where timestamps order is important. But here it's completely different, we won't be displaying/processing activity's timestamp in any other timezone than it's local one (aside from import).

**Comment by laufhannes** [comment]
Well, @mipapo and me needed some time in a private chat to get what we were talking about. And I guess we now want the same thing but use different terms. The above is how we understood stackoverflow's advices and what we now called 'store in utc', but I think it does not matter how we call it.

If the current time is `2015-10-07 10:25 CEST`, this is `2015-10-07 08:25 UTC = 1444206300`. That is what's currently saved in our database and we need the information 'CEST' to get the right time back. What we now want is to forget 'CEST' and save `2015-10-07 10:25 UTC = 1444213500` so that we never ever have to care about the time zone of that activity.

**Comment by grzeg1** [comment]
That's exactly what I meant ;)

**Comment by mipapo** [comment]
````Per-connection time zones. Each client that connects has its own time zone setting, given by the session time_zone variable. Initially, the session variable takes its value from the global time_zone variable, but the client can change its own time zone with this statement. ````

I push the branch as soon as I'm ready.

**Comment by mipapo** [comment]
@grzeg1 @laufhannes 
Because I have a different view to this topic I wanna discuss this again:
When I have a look to my private calendar in google I see my schedules in my timezone. Google saves them in UTC. 
Now I'm travelling to America/NYC: I set my timezone in my calendar to America/NYC. Why? Because I wanna know when my schedules start in the time of my current timezone. 
If we would do the same in RUNALYZE there are to problems. Statistics with night/day information would may show wrong statistics, because runs in another timezone could shown up as done in the night. To show up this correct this would require to store the timezone for each activity.

I don't think the table from @laufhannes is correct or does cover all cases. I'm talking about "Store in UTC":
If we would store everything in UTC you would set `date_default_timezone_set` to the timezone, which has been set by the user. 
Every activity, every body value, every comment (if we would have a community) would be saved in UTC in the database. 
When you don't set  `date_default_timezone_set` to the users timezone and leave it at UTC-+0: 
 * comments of users, registration date, last login date would be shown in UTC always (and that would be wrong)

`The only need for user's time zone is to define NOW.`
Is do define NOW and show up other users content in the correct time zone.

**Comment by laufhannes** [comment]
> Google saves them in UTC.

No. If google would save them in UTC and would ignore all timezones, you would not need to change your timezone to keep the dates.

Again: "*Store in UTC*" means in this discussion "*Forget about timezone. If it's at 5pm, it's at 5pm.*", see https://github.com/Runalyze/Runalyze/issues/1145#issuecomment-146121222.

> If we would store everything in UTC you would set date_default_timezone_set to the timezone, which has been set by the user.

No. Time zone is set to *+00:00* to display activity timestamps.

> comments of users, registration date, last login date would be shown in UTC always (and that would be wrong)

The whole discussion is about timestamps of activities - not any other timestamps. Comments (which are a few years away) and other general dates need a correct timestamp when reading/saving - right.

**Comment by mipapo** [comment]
Google saves them in UTC. When you use the google calendar online you can switch in another timezone. 

> No. Time zone is set to +00:00 to display activity timestamps. 
Okay, if you and @grzeg1 think that way we should to it that way. When 5pm is 5pm.
But `date_default_timezone_set` should be set to the users timezone and just for the activities this setting should be ignored.

**Comment by mipapo** [comment]
Okay, although I see it in some cases different, I accept that 5pm is 5pm. 

One last thing: When I do a query like "Give me all activities from day x". There you would need UTC

**Comment by laufhannes** [comment]
> One last thing: When I do a query like "Give me all activities from day x". There you would need UTC

Not sure what you mean by "*There you would need UTC*". For activities, we always forget about any timezone:
```sql
SELECT * FROM `runalyze_training` WHERE `time` BETWEEN "'.strtotime('20.12.2015 00:00 UTC').'" AND "'.strtotime('20.12.2015 23:59 UTC').'"
```

I don't care what we use as "default". Maybe it's right to keep php/mysql at server time zone and only forget about time zone when handling with activity's timestamps. One drawback: `YEAR(FROM_UNIXTIME())` won't work for activities anymore.

**Comment by mipapo** [comment]
@grzeg1 FYI -  we got it now ;)

**Comment by grzeg1** [comment]
Two things come to my mind:
- whatever way dates are stored, we should store time zone info for them (possibly in a separate field), this field can be then used to re-calculate absolute timestamp or perform migration to a different way of storing timestamps
- all unix timestamps should be migrated to datetime/timestamp data types

**Comment by mipapo** [comment]
- [ ] For now just the timezone information for the activities are missing. 
- [ ] And there will be some places where the correct time zone is missing.
- [ ] A refactor script for the timezone is missing

**Comment by mipapo** [comment]
We decided to stay with unix timestamps. 
@grzeg1  For storing the timezone information on a separate field I'm not sure if it's the best way to just store the offset in minutes. I think it would be okay. Calculations for any statistics with timezone information would be easy to realize. (Seconds would be fine too, but are overkill).
